### ISSUE TITLE
prterun error 'insufficient slots' with --map-by ppr:1:core option

### ISSUE THREAD

**Issue by drwootton** [openpmix/prrte#666]
# Background information
## What version of the PMIx Reference Server are you using? (e.g., v1.0, v2.1, git master @ hash, etc.)

Master branch
## What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)

Master branch
## Please describe the system on which you are running

-    Operating system/version:

    RHEL 7.7

-   Computer hardware:

4 Power 8 nodes, 2 sockets each with 10 cores (20 total) and 8 hwthreads per core (160 total)

-   Network type:

Hostfile specifies 4 nodes, each with slots=1 keyword
## Details of the problem

I ran the command 
```
prterun -n 4 --hostfile hostfile1 --map-by ppr:1:core taskinfo
```
This failed with an error message 
```

There are not enough slots available in the system to satisfy the 4
slots that were requested by the application:

  taskinfo

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the PRTE term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which PRTE processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, PRTE defaults to the number of processor cores

In all the above cases, if you want PRTE to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
number of available slots when deciding the number of processes to
launch.
```
I expected this command to run to normal completion since I have 4 slots total across 4 nodes in the cluster

I also tried to run 
```
 prterun -n 32  --hostfile hostfile8 --map-by ppr:8:hwthread taskinfo
```
and got the same error message, where I expected the program to run to normal completion.

**Comment by jjhursey** [comment]
Playing with this a bit, I agree that the `slots=` value in the hostfile is conflicting with the `ppr` value in the `--map-by`. The `--map-by ppr:1:core` assumes that the hostfile either contains no `slots=` argument or that `slots=N` where `N` is the number of cores on the machine.

Or, more generally `--map-by ppr:N:OBJ` requires that `slots=M` where `M = N * Num_OBJ_on_node` and `-np` (if specified) be less than or equal to the sum of M across all nodes.

My (maybe incorrect) assumption is that `prterun -n 4 --hostfile hostfile1 --map-by ppr:1:core taskinfo` would map 1 process to the first core on each of the four nodes since after mapping one process to a node it reaches its slot limit and should move to the next node. But maybe I am applying round-robin semantics to the PPR mapper when I shouldn't.

**Comment by rhc54** [comment]
> Or, more generally --map-by ppr:N:OBJ requires that slots=M where M = N * Num_OBJ_on_node and -np (if specified) be less than or equal to the sum of M across all nodes.

This is correct - or you have to set OVERSUBSCRIBE

> prterun -n 4 --hostfile hostfile1 --map-by ppr:1:core taskinfo would map 1 process to the first core on each of the four nodes since after mapping one process to a node it reaches its slot limit and should move to the next node

I'm not sure I understand the scenario here. Does your hostfile specify `slots=1`? If so, then yes, your assumption would be correct. If there is no `slots=M` in the hostfile, then we would detect the number of cores on the node and fill up to that number, constrained by `-np`.

**Comment by jjhursey** [comment]
Yes in this scenario `hostfile1` contains:
```
c712f6n01 slots=1
c712f6n02 slots=1
c712f6n03 slots=1
c712f6n04 slots=1
```

If I run the following I get an error:
```
$$ prterun -n 4 --hostfile hostfile1 --map-by ppr:1:core hostname
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 4
slots that were requested by the application:

  hostname

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the PRTE term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which PRTE processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, PRTE defaults to the number of processor cores

In all the above cases, if you want PRTE to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
number of available slots when deciding the number of processes to
launch.
--------------------------------------------------------------------------
```

Where I would expect the output to be one process on the first core of each of the four nodes.

It sounds like this is a bug that needs to be investigated.

**Comment by jjhursey** [comment]
Fixed by PR #713 Thanks!
